packages/pds/src/pipethrough.ts

```diff
@@ -145,18 +92,70 @@ export const createUrlAndHeaders = async (
   return { url, headers }
 }
 
-export const doProxy = async (url: URL, reqInit: RequestInit) => {
+const formatReqInit = (
+  req: express.Request,
+  headers: Record<string, string>,
+  body?: Uint8Array | ReadableStream,
+): RequestInit => {
+  if (req.method === 'GET') {
+    return {
+      method: 'get',
+      headers,
+    }
+  } else if (req.method === 'HEAD') {
+    return {
+      method: 'head',
+      headers,
+    }
+  } else if (req.method === 'POST') {
+    return {
+      method: 'post',
+      headers,
+      body,
+      duplex: 'half',
+    } as RequestInit
+  } else {
+    throw new InvalidRequestError('Method not found')
+  }
+}
+
+export const parseProxyHeader = async (
+  ctx: AppContext,
+  req: express.Request,
+): Promise<{ did: string; serviceUrl: string } | undefined> => {
+  const proxyTo = req.header('atproto-proxy')
+  if (!proxyTo) return
+  const [did, serviceId] = proxyTo.split('#')
+  if (!serviceId) {
+    throw new InvalidRequestError('no service id specified')
+  }
+  const didDoc = await ctx.idResolver.did.resolve(did)
+  if (!didDoc) {
+    throw new InvalidRequestError('could not resolve proxy did')
+  }
+  const serviceUrl = getServiceEndpoint(didDoc, { id: `#${serviceId}` })
+  if (!serviceUrl) {
+    throw new InvalidRequestError('could not resolve proxy did service url')
+  }
+  return { did, serviceUrl }
+}
+
+// Sending request
+// -------------------
+
+export const makeRequest = async (
+  url: URL,
+  reqInit: RequestInit,
+): Promise<Response> => {
   let res: Response
-  let buffer: ArrayBuffer
   try {
     res = await fetch(url, reqInit)
```

Comment:
Something to look out for in general is ensuring we always read the request body.  By not reading the body, I think we'd end-up leaking connections and other resources.

Reply:
I think we're all set on this, but just wanted to put it out there!

Reply:
yup good looks 👍 

think we always do read it - but taking a look back at it, i'm gonna adjust the error handling just a smidge 👌